panic: runtime error: index out of range [0] with length 0 on PipelineLoop custom tasks

# Expected Behavior

Tekton Pipelines Controller should skip Runs that are indirectly created by the PipelineRun (e.g. by custom tasks).

# Actual Behavior

Tekton Pipelines Controller panics with:

```
panic: runtime error: index out of range [0] with length 0

goroutine 250 [running]:
github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun.updatePipelineRunStatusFromRuns(0xc000e78060, 0xc000cfa000, {0xc000d12010, 0x1, 0x0})
	github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/pipelinerun.go:1278 +0x2b1
github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun.(*Reconciler).updatePipelineRunStatusFromInformer(0xc0004946e0, {0x21b2f08, 0xc000c580c0}, 0xc000cfa000)
	github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/pipelinerun.go:1171 +0x288
github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun.(*Reconciler).ReconcileKind(0xc0004946e0, {0x21b2f08, 0xc0010180f0}, 0xc000cfa000)
	github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/pipelinerun.go:218 +0x11f7
github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/pipelinerun.(*reconcilerImpl).Reconcile(0xc00032ebe0, {0x21b2f08, 0xc000a39650}, {0xc0008a1440, 0x20})
	github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/pipelinerun/reconciler.go:235 +0x6a2
github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller.(*Impl).processNextWorkItem(0xc0004d22a0)
	github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:542 +0x5a5
github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller.(*Impl).RunContext.func3()
	github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:478 +0x68
created by github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller.(*Impl).RunContext
	github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:476 +0x2de
kubectl -n tekton-pipelines get deployment tekton-pipelines-controller -o yaml | grep image
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app.kubernetes.io/component":"controller","app.kubernetes.io/instance":"default","app.kubernetes.io/name":"controller","app.kubernetes.io/part-of":"tekton-pipelines","app.kubernetes.io/version":"v0.33.2","pipeline.tekton.dev/release":"v0.33.2","version":"v0.33.2"},"name":"tekton-pipelines-controller","namespace":"tekton-pipelines"},"spec":{"replicas":1,"selector":{"matchLabels":{"app.kubernetes.io/c
```

# Steps to Reproduce the Problem

1. Enable custom tasks feature flag
2. Deploy https://github.com/kubeflow/kfp-tekton/tree/master/tekton-catalog/pipeline-loops custom controller
3. Run example: https://github.com/kubeflow/kfp-tekton/blob/master/tekton-catalog/pipeline-loops/examples/loop-example-with-parallelism.yaml

# Additional Info

- Kubernetes version:

```
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-02-16T12:30:48Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:34:54Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
  ```

- Tekton Pipeline version: v0.33.2

--- a/pkg/reconciler/pipelinerun/pipelinerun.go
+++ b/pkg/reconciler/pipelinerun/pipelinerun.go
@@ -1297,7 +1297,7 @@ func updatePipelineRunStatusFromRuns(logger *zap.SugaredLogger, pr *v1beta1.Pipe
 	for _, run := range runs {
 		// Only process Runs that are owned by this PipelineRun.
 		// This skips Runs that are indirectly created by the PipelineRun (e.g. by custom tasks).
-		if len(run.OwnerReferences) < 1 && run.OwnerReferences[0].UID != pr.ObjectMeta.UID {
+		if len(run.OwnerReferences) < 1 || run.OwnerReferences[0].UID != pr.ObjectMeta.UID {
 			logger.Debugf("Found a Run %s that is not owned by this PipelineRun", run.Name)
 			continue
 		}

--- a/pkg/reconciler/pipelinerun/pipelinerun_test.go
+++ b/pkg/reconciler/pipelinerun/pipelinerun_test.go
@@ -6776,6 +6776,17 @@ func TestUpdatePipelineRunStatusFromTaskRuns(t *testing.T) {
 		},
 	}
 
+	taskRunsWithNoOwner := []*v1beta1.TaskRun{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pr-task-1-xxyyy",
+				Labels: map[string]string{
+					pipeline.PipelineTaskLabelKey: "task-1",
+				},
+			},
+		},
+	}
+
 	tcs := []struct {
 		prName           string
 		prStatus         v1beta1.PipelineRunStatus
@@ -6848,6 +6859,11 @@ func TestUpdatePipelineRunStatusFromTaskRuns(t *testing.T) {
 			prStatus:         prStatusWithEmptyTaskRuns,
 			trs:              taskRunsFromAnotherPR,
 			expectedPrStatus: prStatusWithEmptyTaskRuns,
+		}, {
+			prName:           "tr-with-no-owner",
+			prStatus:         prStatusWithEmptyTaskRuns,
+			trs:              taskRunsWithNoOwner,
+			expectedPrStatus: prStatusWithEmptyTaskRuns,
 		},
 	}
 
@@ -6875,6 +6891,211 @@ func TestUpdatePipelineRunStatusFromTaskRuns(t *testing.T) {
 				}
 				actualPrStatus.PipelineRunStatusFields.TaskRuns = fixedTaskRuns
 			}
+
+			if d := cmp.Diff(tc.expectedPrStatus, actualPrStatus); d != "" {
+				t.Errorf("expected the PipelineRun status to match %#v. Diff %s", tc.expectedPrStatus, diff.PrintWantGot(d))
+			}
+		})
+	}
+}
+
+func TestUpdatePipelineRunStatusFromRuns(t *testing.T) {
+
+	prUID := types.UID("11111111-1111-1111-1111-111111111111")
+	otherPrUID := types.UID("22222222-2222-2222-2222-222222222222")
+
+	prRunningStatus := duckv1beta1.Status{
+		Conditions: []apis.Condition{
+			{
+				Type:    "Succeeded",
+				Status:  "Unknown",
+				Reason:  "Running",
+				Message: "Not all Tasks in the Pipeline have finished executing",
+			},
+		},
+	}
+
+	prStatusWithSomeRuns := v1beta1.PipelineRunStatus{
+		Status: prRunningStatus,
+		PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+			Runs: map[string]*v1beta1.PipelineRunRunStatus{
+				"pr-run-1-xxyyy": {
+					PipelineTaskName: "run-1",
+					Status:           &v1alpha1.RunStatus{},
+				},
+				"pr-run-2-xxyyy": {
+					PipelineTaskName: "run-2",
+					Status:           nil,
+				},
+			},
+		},
+	}
+
+	prStatusWithAllRuns := v1beta1.PipelineRunStatus{
+		Status: prRunningStatus,
+		PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+			Runs: map[string]*v1beta1.PipelineRunRunStatus{
+				"pr-run-1-xxyyy": {
+					PipelineTaskName: "run-1",
+					Status:           &v1alpha1.RunStatus{},
+				},
+				"pr-run-2-xxyyy": {
+					PipelineTaskName: "run-2",
+					Status:           nil,
+				},
+				"pr-run-3-xxyyy": {
+					PipelineTaskName: "run-3",
+					Status:           &v1alpha1.RunStatus{},
+				},
+			},
+		},
+	}
+
+	prStatusWithEmptyRuns := v1beta1.PipelineRunStatus{
+		Status: prRunningStatus,
+		PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+			Runs: map[string]*v1beta1.PipelineRunRunStatus{},
+		},
+	}
+
+	prStatusRecoveredSimple := v1beta1.PipelineRunStatus{
+		Status: prRunningStatus,
+		PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+			Runs: map[string]*v1beta1.PipelineRunRunStatus{
+				"pr-run-1-xxyyy": {
+					PipelineTaskName: "run-1",
+					Status:           &v1alpha1.RunStatus{},
+				},
+			},
+		},
+	}
+
+	allRuns := []*v1alpha1.Run{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pr-run-1-xxyyy",
+				Labels: map[string]string{
+					pipeline.PipelineTaskLabelKey: "run-1",
+				},
+				OwnerReferences: []metav1.OwnerReference{{UID: prUID}},
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pr-run-2-xxyyy",
+				Labels: map[string]string{
+					pipeline.PipelineTaskLabelKey: "run-2",
+				},
+				OwnerReferences: []metav1.OwnerReference{{UID: prUID}},
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pr-run-3-xxyyy",
+				Labels: map[string]string{
+					pipeline.PipelineTaskLabelKey: "run-3",
+				},
+				OwnerReferences: []metav1.OwnerReference{{UID: prUID}},
+			},
+		},
+	}
+
+	runsFromAnotherPR := []*v1alpha1.Run{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pr-run-1-xxyyy",
+				Labels: map[string]string{
+					pipeline.PipelineTaskLabelKey: "run-1",
+				},
+				OwnerReferences: []metav1.OwnerReference{{UID: otherPrUID}},
+			},
+		},
+	}
+
+	runsWithNoOwner := []*v1alpha1.Run{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "pr-run-1-xxyyy",
+				Labels: map[string]string{
+					pipeline.PipelineTaskLabelKey: "run-1",
+				},
+			},
+		},
+	}
+
+	tcs := []struct {
+		prName           string
+		prStatus         v1beta1.PipelineRunStatus
+		runs             []*v1alpha1.Run
+		expectedPrStatus v1beta1.PipelineRunStatus
+	}{
+		{
+			prName:           "no-status-no-runs",
+			prStatus:         v1beta1.PipelineRunStatus{},
+			runs:             nil,
+			expectedPrStatus: v1beta1.PipelineRunStatus{},
+		}, {
+			prName:           "status-no-runs",
+			prStatus:         prStatusWithSomeRuns,
+			runs:             nil,
+			expectedPrStatus: prStatusWithSomeRuns,
+		}, {
+			prName:   "status-nil-runs",
+			prStatus: prStatusWithEmptyRuns,
+			runs: []*v1alpha1.Run{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "pr-run-1-xxyyy",
+						Labels: map[string]string{
+							pipeline.PipelineTaskLabelKey: "run-1",
+						},
+						OwnerReferences: []metav1.OwnerReference{{UID: prUID}},
+					},
+				},
+			},
+			expectedPrStatus: prStatusRecoveredSimple,
+		}, {
+			prName:   "status-missing-runs",
+			prStatus: prStatusWithSomeRuns,
+			runs: []*v1alpha1.Run{{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "pr-run-3-xxyyy",
+					Labels: map[string]string{
+						pipeline.PipelineTaskLabelKey: "run-3",
+					},
+					OwnerReferences: []metav1.OwnerReference{{UID: prUID}},
+				},
+			}},
+			expectedPrStatus: prStatusWithAllRuns,
+		}, {
+			prName:           "status-matching-runs-pr",
+			prStatus:         prStatusWithAllRuns,
+			runs:             allRuns,
+			expectedPrStatus: prStatusWithAllRuns,
+		}, {
+			prName:           "run-from-another-pr",
+			prStatus:         prStatusWithEmptyRuns,
+			runs:             runsFromAnotherPR,
+			expectedPrStatus: prStatusWithEmptyRuns,
+		}, {
+			prName:           "run-with-no-owner",
+			prStatus:         prStatusWithEmptyRuns,
+			runs:             runsWithNoOwner,
+			expectedPrStatus: prStatusWithEmptyRuns,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.prName, func(t *testing.T) {
+			logger := logtesting.TestLogger(t)
+
+			pr := &v1beta1.PipelineRun{
+				ObjectMeta: metav1.ObjectMeta{Name: tc.prName, UID: prUID},
+				Status:     tc.prStatus,
+			}
+
+			updatePipelineRunStatusFromRuns(logger, pr, tc.runs)
+			actualPrStatus := pr.Status
 
 			if d := cmp.Diff(tc.expectedPrStatus, actualPrStatus); d != "" {
 				t.Errorf("expected the PipelineRun status to match %#v. Diff %s", tc.expectedPrStatus, diff.PrintWantGot(d))

